### ISSUE TITLE
Verify library detection

### ISSUE THREAD

**Issue by rviscomi** [HTTPArchive/httparchive#93]
https://github.com/HTTPArchive/httparchive/issues/77 (library detection) goes live in the upcoming 4/15 run. Verify the data to ensure that the detection is working properly. For example, take a random sample and manually inspect the pages for inconsistencies.

**Comment by rviscomi** [comment]
The 4/15 HAR data is finally available in BigQuery. It's JSON encoded, which I thought would help analysis but it actually turns out to just get in the way. If the detection results were just returned as plain JS objects, they'd be fully queryable with the `JSON_EXTRACT` function. Methinks I should remove the `JSON.stringify` and look into postprocessing to fix up the 4/15 and 5/1 runs.

To get started with the data already available, this UDF helps with "unstringifying" the array of library objects and converts it into a repeated field.

```
CREATE TEMPORARY FUNCTION parseJson(libs STRING)
RETURNS ARRAY<STRUCT<name STRING, version STRING>>
LANGUAGE js AS """
  return JSON.parse(libs);
""";
```

Example usage:

```
SELECT
    parseJson(TRIM(JSON_EXTRACT(payload, "$._third-parties"), '"'))
FROM
    `httparchive.har.2017_04_15_chrome_pages`
LIMIT
    1
```

It should yield something like this: `"[{\"name\":\"jQuery\",\"version\":\"1.5.2\"}]"`.

Note that many (~2000) pages actually have _double stringification_. For example: `"\"[{\\\"name\\\":\\\"MooTools\\\",\\\"version\\\":\\\"1.2.4\\\"}]\""`. No idea how this happens but removing the stringify call in the custom metric should avoid this entirely.


Working on a full-scale query over the entire table here: https://bigquery.cloud.google.com/savedquery/226352634162:d42f2526767b465d98b63e62591f93a5

**Comment by igrigorik** [comment]
> If the detection results were just returned as plain JS objects, they'd be fully queryable with the JSON_EXTRACT function.

Does WPT custom metrics support that? I was under impression that it has to be a string? If that's the case, one hack is to add some special case logic to "deserialize" some custom metrics as part of our dataflow transform?

**Comment by rviscomi** [comment]
Ah yeah that's right.

Closing the loop, I managed to get a relatively clean query that ignores any invalid formatting:

```
CREATE TEMPORARY FUNCTION parseJson(libs STRING)
RETURNS ARRAY<STRUCT<name STRING, version STRING>>
LANGUAGE js AS """
  try {
    return JSON.parse(libs);
  } catch (e) {
    return [];
  }
""";

SELECT
  url,
  lib
FROM (
  SELECT
    url,
    parseJson(TRIM(JSON_EXTRACT(payload, "$._third-parties"), '"')) AS libs
  FROM
    `httparchive.har.2017_04_15_chrome_pages`)
CROSS JOIN
  UNNEST(libs) AS lib
ORDER BY
  url
```

The secret sauce is just adding try/catch around the JSON parsing. So the ~2000 bad pages will be ignored. Saved the results to https://bigquery.cloud.google.com/table/httparchive:scratchspace.2017_04_15_js_libs

I wrote up a post on the HA forum to demonstrate how to work with the data: https://discuss.httparchive.org/t/javascript-library-detection/955

On the validation front, one sample query is to measure the usage of jQuery. It came out to 82% of all pages crawled. That corresponds really well with https://trends.builtwith.com/javascript/jQuery which puts usage at 74.6% of the top 100k and 88.8% of the top 1M. We're measuring the top 500k, so 82% lines up nicely.
